parser/src/format/mod.rs

```diff
@@ -140,15 +147,10 @@ fn format_output(
     let decorator = Decorator::from_config(config);
     let mut buffer = io::BufWriter::new(dest);
     let mut record_count = 0u64;
-    for result in output {
-        let mut value = result.map_err(|e| {
-            tracing::warn!(
-                record_count = record_count,
-                "parsing failed after {} records",
-                record_count
-            );
-            e
-        })?;
+    let error_threshold = config.error_threshold.unwrap_or_default();
+
+    for result in ParseErrorBuffer::new(output, error_threshold) {
```

Comment:
Okay, yeah. That should be doable. I think my initial thought was that malformed records are a problem for any file type. Depending on exactly what it is, there might be different error parsing characteristics, but the possibility of bad data remains.

Ex. Given a `jsonl` file where a row is improperly formatted by being pretty printed into the stream:
```jsonl
{"id": 0, "color": "green"}
{
  "id": 1,
  "color": "blue"
}
{"id": 2, "color": "red"}
```
We'd encounter line 0, parse it correctly. Then line 1,2,3,4 would all fail to parse. Then line 5 would succeed. We'd count that "one" malformed document as 4 `ParseError`s, but we *would* be able to skip over it.

My worry here isn't so much that serde is left in an invalid state, but that deeply nested documents might _parse_ correctly but on an inner subset document (though I suspect that would probably fail schema validation).

If we're trying to parse a file as a _single_ json document and that document is malformed, there's no meaningful usage for an error threshold greater than 0. If _the document_ fails to parse, it fails.

Now, all that said, I don't know if we know how many other file formats should handle malformed records yet. I'm fine with moving the error buffer into the csv module.

> it seems like error threshold that really only makes sense for [...] certain error types. A decompression error should always be treated as terminal, for example.

I wasn't sure about treating all types of `ParseError`s as recoverable/skippable, but I didn't try to differeniate them yet. Your decompression example is a perfect case. I can add a check against a whitelist of the `ParseError`s we allow to be skipped.

Reply:
>  I can add a check against a whitelist of the `ParseError`s we allow to be skipped.

Upon further inspection, I punted on this. I believe that `ParseError::Parse(_)` is the only variant that we _can_ skip and it's on the only variant we can even produce at this stage of parsing. It could be worth revisiting if we want to differentiate finer-grain parsing problems, but for now, I think we're probably okay.